packages/babel-traverse/test/scope.js

```diff
@@ -250,4 +271,103 @@ describe("scope", function() {
       });
     });
   });
+
+  describe("duplicate bindings", () => {
+    /*
+     * These tests do not use the parser as the parser has
+     * its own scope tracking and we want to test the scope tracking
+     * of traverse here and see if it handles duplicate bindings correctly
+     */
+    describe("catch", () => {
+      // try {} catch (e) { let e; }
+      const createTryCatch = function(kind) {
+        return t.tryStatement(
+          t.blockStatement([]),
+          t.catchClause(
+            t.identifier("e"),
+            t.blockStatement([
+              t.variableDeclaration(kind, [
+                t.variableDeclarator(t.identifier("e"), t.stringLiteral("1")),
+              ]),
+            ]),
+          ),
+        );
+      };
+      ["let", "const"].forEach(name => {
+        it(name, () => {
+          const ast = createTryCatch(name);
+
+          expect(() => getPath(ast)).toThrowErrorMatchingSnapshot();
+        });
+      });
+
+      it("var", () => {
+        const ast = createTryCatch("var");
+
+        expect(() => getPath(ast)).not.toThrow();
+      });
+    });
+
+    ["let", "const"].forEach(name => {
+      it(`${name} and function in sub scope`, () => {
+        const ast = [
+          t.variableDeclaration(name, [
+            t.variableDeclarator(t.identifier("foo")),
+          ]),
+          t.blockStatement([
+            t.functionDeclaration(
+              t.identifier("foo"),
+              [],
+              t.blockStatement([]),
+            ),
+          ]),
+        ];
+
+        expect(() => getPath(ast)).not.toThrow();
+      });
+    });
+
+    describe("global", () => {
+      ["let", "const"].forEach(name => {
```

Comment:
We should also test `class/function` and `class/var` (and maybe `function/var` to be sure that it doesn't throw)

Reply:
Okay I added a more generic way to define tests.
Turns out we have some problems, unless I made a mistake in the tests.
```
var foo = 1;
class foo {};
```
and
```
class foo {};
var foo = 1;
```
Class and var never throw.

also does not throw:

```js
function foo() {}
const foo = 1;
```

```js
var foo = 1;
const foo = 1;
```

Thats why I disabled the reverse matching in the test.